### ISSUE TITLE
Rework interrupts handling framework

### ISSUE THREAD

**Issue by lumag** [ARM-software/tf-issues#365]
Current interrupt setup has the following drawbacks:
 - It is possible to set the routing model in several places. In fact current code implies that routing model is related to the interrupt type, which makes no sense to me. In the end we just set the routing for FIQ and IRQs in the SCR_EL3.

- There is no API to configure if the interrupt is edge- or level- triggered.

- All G0 and G1S interrupts have to be declared during the GIC setup. This is not quite logical from
software point of view, as interrupt handlers are declared separately.

- There is no API for multiplexing interrupt handling. I.e. we want to handle several interrupts in EL3 firmware. Currently we'd have to handcode main interrupt handling function, calling individual handlers. We are looking for the code that'd allow to register individual interrupts in the framework code. Of course this can be an optional stage.

We have some code, but I'd like to discuss the issue first, before presenting the code.

**Comment by soby-mathew** [comment]
> It is possible to set the routing model in several places. In fact current code implies that routing model is related to the interrupt type, which makes no sense to me. In the end we just set the routing for FIQ and IRQs in the SCR_EL3.

Yes, you are right that the routing model is tied to the interrupt type. Typically in a system, given an interrupt type, it is to be routed in the same manner and hence the association.

> There is no API to configure if the interrupt is edge- or level- triggered.

There is a pending issue #262 for this. We haven't had time to resolve it yet. It would be very useful if you could contribute towards this.

> All G0 and G1S interrupts have to be declared during the GIC setup. This is not quite logical from
software point of view, as interrupt handlers are declared separately.

This is useful from Trusted Firmware point of view because all the "secure" interrupts can be configured at GIC initialization and this information is static. Hence the driver doesn't need to worry about `dynamic` interrupt configuration. 

> There is no API for multiplexing interrupt handling. I.e. we want to handle several interrupts in EL3 firmware. Currently we'd have to handcode main interrupt handling function, calling individual handlers. We are looking for the code that'd allow to register individual interrupts in the framework code. Of course this can be an optional stage.

Yes, the interrupt management framework is more geared towards "interrupt type" management rather than individual interrupts. This is partly due to the fact that there is no reliable way of getting the interrupt ID other than by "acknowledging" the interrupt. Since the main task of IMF is to route interrupts types,  "acknowledging" the interrupt is not ideal.  It is upto the dispatcher to implement framework for this.

HTH,
Best Regards
Soby Mathew
